pipeline/id_minter/src/main/scala/uk/ac/wellcome/platform/idminter/steps/SourceIdentifierScanner.scala

```diff
@@ -0,0 +1,108 @@
+package uk.ac.wellcome.platform.idminter.steps
+
+import grizzled.slf4j.Logging
+import io.circe.Json
+import io.circe.optics.JsonPath.root
+import io.circe.optics.JsonOptics._
+import monocle.function.Plated
+import uk.ac.wellcome.models.Implicits._
+import uk.ac.wellcome.models.work.internal.SourceIdentifier
+import uk.ac.wellcome.platform.idminter.models.Identifier
+
+import scala.annotation.tailrec
+import scala.util.{Failure, Success, Try}
+
+/**
+  * SourceIdentifierScanner provides 2 methods:
+  *
+  * - `scan` takes Json and returns all of the sourceIdentifiers that are in it
+  * - `update` takes Json and a map of (SourceIdentifier -> Identifier) and adds
+  *   a canonicalId field next to sourceIdentifiers, as well as replacing
+  *   `identifiedType` fields with `type` fields of the same value.
+  *
+  */
+object SourceIdentifierScanner extends Logging {
+  def update(inputJson: Json,
+             identifiers: Map[SourceIdentifier, Identifier]): Try[Json] =
+    Try {
+      val updateNode =
+        (updateNodeType _) compose addCanonicalIdToNode(identifiers)
+      // Plated transforms operate on self-similar *children*
+      // so we need to update the root separately
+      val updatedRoot = updateNode(inputJson)
+      Plated.transform[Json](updateNode)(updatedRoot)
+    }
+
+  private def updateNodeType(node: Json): Json =
+    root.identifiedType.json
+      .getOption(node)
+      .flatMap(_.asString)
+      .map { identifiedType =>
+        root.obj.modify { obj =>
+          ("type", Json.fromString(identifiedType)) +:
+            obj.remove("identifiedType")
+        }(node)
+      }
+      .getOrElse(node)
```

Comment:
Should we throw an exception here if we can’t find a field `"identifiedType"`? Is that ever something that could happen in practice?

Reply:
Welcome to the weird world of circe-optics - that's not something that could happen in practice, because `root.identifiedType.json.getOption` is ensuring that it exists

Reply:
I think we’re talking at cross-purposes – `getOption` ensures you either find it or you don’t, so there won’t be an unexpected runtime error – but what if we’re passed a blob of JSON where `"identifiedType"` isn't present? That could surely happen?

Reply:
Yes absolutely! That's what the `getOrElse` at the end is doing - allowing `node`s without an `identifiedType` to pass straight through.

Reply:
I’m struggling to explain what I mean, I think.

When would the ID minter be asked to mint identifiers on a JSON map that *didn't* have an `"identifiedType"` field? I thought we had that everywhere?

Reply:
Ah, I get you now - I don't know whether we have it everywhere, but I do know that the old ID minter code explicitly handled (and didn't throw an exception for) the case where a thing didn't have an `identifiedType`, so I chose to maintain existing behaviour here.